domain/port/state/state.go

```diff
@@ -82,6 +82,52 @@ WHERE unit_endpoint.unit_uuid = $unitUUID.unit_uuid
 	return groupedPortRanges, nil
 }
 
+// GetAllOpenedPorts returns the opened ports in the model, grouped by unit name.
+//
+// NOTE: We do not group by endpoint here. It is not needed. Instead, we just
+// group by unit name
+func (s *State) GetAllOpenedPorts(ctx context.Context) (network.GroupedPortRanges, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, jujuerrors.Trace(err)
+	}
+
+	query, err := s.Prepare(`
+SELECT DISTINCT &unitNamePortRange.*
```

Comment:
Is it possible to have duplicates in the DB? Is this not something we can enforce on the way into the DB instead? Distinct can be expensive, I'm sure in this case its fine because you're not going to have many ports, but still.

Reply:
So we will need to de-dup nonetheless. So best to do in the db